hail/python/hail/nd/nd.py

```diff
@@ -119,4 +120,41 @@ def ones(shape, dtype=hl.tfloat64):
        :class:`.NDArrayNumericExpression`
            NDArray of the specified size full of ones.
        """
-    return full(shape, 1, dtype)
\ No newline at end of file
+    return full(shape, 1, dtype)
+
+
+@typecheck(nd=expr_ndarray(), mode=str)
+def qr(nd, mode="reduced"):
+    """Performs a QR decomposition.
+
+    :param nd: A 2 dimensional ndarray, shape(M, N)
+    :param mode: One of "reduced", "complete", "r", or "raw"
+        If K = min(M, N), then
+        - `reduced`: returns q and r with dimensions (M, K), (K, N)
+        - `complete`: returns q and r with dimensions (M, M), (M, N)
+        - `r`: returns only r with dimensions (K, N)
+        - `raw`: returns h, tau with dimensions (N, M), (K,)
```

Comment:
Yeah, this would be cool, but like you mentioned I don't think it's possible to resolve `reduced` and `complete`. We could only do this if we had matrices that have their shape as part of their type (which we don't want), or if we did something like `qr(...).complete_r` vs `qr(...).r` or equivalent.